drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1453 @@
+            DPRINT1("SOCK_RAW not supported\n");
+            status = STATUS_NOT_SUPPORTED;
+            goto err_out;
+
+        default:
+            status = STATUS_INVALID_PARAMETER;
+            goto err_out;
+    }
+
+    s = ExAllocatePoolWithTag(NonPagedPool, sizeof(*s), TAG_NETIO);
+    if (s == NULL)
+    {
+        DPRINT1("WskSocket: Out of memory\n");
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto err_out;
+    }
+    s->family = AddressFamily;
+    s->type = SocketType;
+    s->proto = Protocol;
+    s->WskFlags = Flags;
+    s->user_context = SocketContext;
+    s->LocalAddressHandle = NULL;
+    s->LocalAddressFile = NULL;
+    s->Flags = 0;
+    s->ListenDispatch = Dispatch;
+    s->RefCount = 1;            /* SocketPut() is in WskCloseSocket */
+    s->ConnectionHandle = NULL;
+    s->ConnectionFile = NULL;
+    s->ConnectionFileAssociated = FALSE;
+    s->ListenIrp = NULL;
+    s->ListenThreadHandle = NULL;
+    s->ListenThreadShouldRun = FALSE;
+    s->ListenCancelled = FALSE;
+    s->ListenSocket = NULL;
+    RtlZeroMemory(&s->LocalAddress, sizeof(s->LocalAddress));
+    RtlZeroMemory(&s->RemoteAddress, sizeof(s->RemoteAddress));
+    s->NextSocketToPut = NULL;
+    s->NumSocketPuts = 0;
+
+    switch (SocketType)
+    {
+        case SOCK_DGRAM:
+            s->s.Dispatch = &UdpDispatch;
+            RtlInitUnicodeString(&s->TdiName, L"\\Device\\Udp");
+            break;
+        case SOCK_STREAM:
+            s->s.Dispatch = &TcpDispatch;
+            RtlInitUnicodeString(&s->TdiName, L"\\Device\\Tcp");
+
+            if (Flags != WSK_FLAG_LISTEN_SOCKET)
+            {
+                status = TdiOpenConnectionEndpointFile(&s->TdiName, &s->ConnectionHandle, &s->ConnectionFile);
+                if (status != STATUS_SUCCESS)
+                {
+                    DPRINT1("Could not open TDI handle, status is %x\n", status);
+                    ExFreePoolWithTag(s, TAG_NETIO);
+                    goto err_out;
+                }
+            }
+            else
+            {
+                KeInitializeEvent(&s->StartListenEvent, SynchronizationEvent, FALSE);
+                s->ListenThreadShouldRun = TRUE;
+
+                status = PsCreateSystemThread(&s->ListenThreadHandle, THREAD_ALL_ACCESS, NULL, NULL, NULL, RequeueListenThread, s);
+                if (status != STATUS_SUCCESS)
+                {
+                    DPRINT1("Could not start listen thread, status is %x\n", status);
+                    ExFreePoolWithTag(s, TAG_NETIO);
+                    goto err_out;
+                }
+                status = ObReferenceObjectByHandle(s->ListenThreadHandle, THREAD_ALL_ACCESS, NULL, KernelMode, (void **)&s->ListenThread, NULL);
+                if (status != STATUS_SUCCESS)
+                {
+                    DPRINT1("Could not get a PKTHREAD object, status is %x\n", status);
+                    ExFreePoolWithTag(s, TAG_NETIO);
+                    goto err_out;
+                }
```

Comment:
See also https://www-user.tu-chemnitz.de/~heha/oney_wdm/ch09e.htm
```suggestion
                }
                ZwClose(s->ListenThreadHandle);
```

Reply:
Ok thanks applied.